File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/Observer/concrete_observer.go
package Observer

import "fmt"
import "sync"

// LoginObserver 登录观察者
type LoginObserver struct{}

var loginObserverInstance *LoginObserver
var loginObserverOnce sync.Once

// GetLoginObserver 获取登录观察者实例
func GetLoginObserver() *LoginObserver {
	loginObserverOnce.Do(func() {
		loginObserverInstance = &LoginObserver{}
	})
	return loginObserverInstance
}

// Update 更新
func (lo *LoginObserver) Update(subject ISubject) {
	if s, ok := subject.(*Subject); ok {
		GetAccountData().LoginAccount(s.GetState(0), s.GetState(1))
	}
}

// RegisterObserver 注册观察者
type RegisterObserver struct{}

var registerObserverInstance *RegisterObserver
var registerObserverOnce sync.Once

// GetRegisterObserver 获取注册观察者实例
func GetRegisterObserver() *RegisterObserver {
	registerObserverOnce.Do(func() {
		registerObserverInstance = &RegisterObserver{}
	})
	return registerObserverInstance
}

// Update 更新
func (ro *RegisterObserver) Update(subject ISubject) {
	if s, ok := subject.(*Subject); ok {
		GetAccountData().RegisterAccount(s.GetState(0), s.GetState(1))
	}
}

// LoginSuccessful 登录成功观察者
type LoginSuccessful struct{}

var loginSuccessfulInstance *LoginSuccessful
var loginSuccessfulOnce sync.Once

// GetLoginSuccessfulObserver 获取登录成功观察者实例
func GetLoginSuccessfulObserver() *LoginSuccessful {
	loginSuccessfulOnce.Do(func() {
		loginSuccessfulInstance = &LoginSuccessful{}
	})
	return loginSuccessfulInstance
}

// Update 更新
func (ls *LoginSuccessful) Update(subject ISubject) {
	fmt.Println("Login Successful")
}

// LoginFailed 登录失败观察者
type LoginFailed struct{}

var loginFailedInstance *LoginFailed
var loginFailedOnce sync.Once

// GetLoginFailedObserver 获取登录失败观察者实例
func GetLoginFailedObserver() *LoginFailed {
	loginFailedOnce.Do(func() {
		loginFailedInstance = &LoginFailed{}
	})
	return loginFailedInstance
}

// Update 更新
func (lf *LoginFailed) Update(subject ISubject) {
	fmt.Println("Login Failed")
}

// RegisterSuccessful 注册成功观察者
type RegisterSuccessful struct{}

var registerSuccessfulInstance *RegisterSuccessful
var registerSuccessfulOnce sync.Once

// GetRegisterSuccessfulObserver 获取注册成功观察者实例
func GetRegisterSuccessfulObserver() *RegisterSuccessful {
	registerSuccessfulOnce.Do(func() {
		registerSuccessfulInstance = &RegisterSuccessful{}
	})
	return registerSuccessfulInstance
}

// Update 更新
func (rs *RegisterSuccessful) Update(subject ISubject) {
	fmt.Println("Register-Account Successful")
}

// RegisterFailed 注册失败观察者
type RegisterFailed struct{}

var registerFailedInstance *RegisterFailed
var registerFailedOnce sync.Once

// GetRegisterFailedObserver 获取注册失败观察者实例
func GetRegisterFailedObserver() *RegisterFailed {
	registerFailedOnce.Do(func() {
		registerFailedInstance = &RegisterFailed{}
	})
	return registerFailedInstance
}

// Update 更新
func (rf *RegisterFailed) Update(subject ISubject) {
	fmt.Println("Register-Account Failed")
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Observer/ConcreteObserver.lua
local ObserverModule = require("Observer")
local IObserver = ObserverModule.IObserver

-- 登录观察者
local LoginObserver = {}
LoginObserver.__index = LoginObserver
setmetatable(LoginObserver, IObserver)

local loginObserverInstance = nil

function LoginObserver:getInstance()
    if not loginObserverInstance then
        loginObserverInstance = LoginObserver:new()
    end
    return loginObserverInstance
end

function LoginObserver:new()
    local o = IObserver:new()
    setmetatable(o, self)
    return o
end

function LoginObserver:Update(subject)
    local AccountData = require("AccountData")
    if subject.GetState then
        AccountData.getInstance():LoginAccount(subject:GetState(0), subject:GetState(1))
    end
end

-- 注册观察者
local RegisterObserver = {}
RegisterObserver.__index = RegisterObserver
setmetatable(RegisterObserver, IObserver)

local registerObserverInstance = nil

function RegisterObserver:getInstance()
    if not registerObserverInstance then
        registerObserverInstance = RegisterObserver:new()
    end
    return registerObserverInstance
end

function RegisterObserver:new()
    local o = IObserver:new()
    setmetatable(o, self)
    return o
end

function RegisterObserver:Update(subject)
    local AccountData = require("AccountData")
    if subject.GetState then
        AccountData.getInstance():RegisterAccount(subject:GetState(0), subject:GetState(1))
    end
end

-- 登录成功观察者
local LoginSuccessful = {}
LoginSuccessful.__index = LoginSuccessful
setmetatable(LoginSuccessful, IObserver)

local loginSuccessfulInstance = nil

function LoginSuccessful:getInstance()
    if not loginSuccessfulInstance then
        loginSuccessfulInstance = LoginSuccessful:new()
    end
    return loginSuccessfulInstance
end

function LoginSuccessful:new()
    local o = IObserver:new()
    setmetatable(o, self)
    return o
end

function LoginSuccessful:Update(subject)
    print("Login Successful")
end

-- 登录失败观察者
local LoginFailed = {}
LoginFailed.__index = LoginFailed
setmetatable(LoginFailed, IObserver)

local loginFailedInstance = nil

function LoginFailed:getInstance()
    if not loginFailedInstance then
        loginFailedInstance = LoginFailed:new()
    end
    return loginFailedInstance
end

function LoginFailed:new()
    local o = IObserver:new()
    setmetatable(o, self)
    return o
end

function LoginFailed:Update(subject)
    print("Login Failed")
end

-- 注册成功观察者
local RegisterSuccessful = {}
RegisterSuccessful.__index = RegisterSuccessful
setmetatable(RegisterSuccessful, IObserver)

local registerSuccessfulInstance = nil

function RegisterSuccessful:getInstance()
    if not registerSuccessfulInstance then
        registerSuccessfulInstance = RegisterSuccessful:new()
    end
    return registerSuccessfulInstance
end

function RegisterSuccessful:new()
    local o = IObserver:new()
    setmetatable(o, self)
    return o
end

function RegisterSuccessful:Update(subject)
    print("Register-Account Successful")
end

-- 注册失败观察者
local RegisterFailed = {}
RegisterFailed.__index = RegisterFailed
setmetatable(RegisterFailed, IObserver)

local registerFailedInstance = nil

function RegisterFailed:getInstance()
    if not registerFailedInstance then
        registerFailedInstance = RegisterFailed:new()
    end
    return registerFailedInstance
end

function RegisterFailed:new()
    local o = IObserver:new()
    setmetatable(o, self)
    return o
end

function RegisterFailed:Update(subject)
    print("Register-Account Failed")
end

-- 导出函数
return {
    getLoginObserver = function() return LoginObserver:getInstance() end,
    getRegisterObserver = function() return RegisterObserver:getInstance() end,
    getLoginSuccessfulObserver = function() return LoginSuccessful:getInstance() end,
    getLoginFailedObserver = function() return LoginFailed:getInstance() end,
    getRegisterSuccessfulObserver = function() return RegisterSuccessful:getInstance() end,
    getRegisterFailedObserver = function() return RegisterFailed:getInstance() end
}
